### PR TITLE
Mob properties

### PR DESCRIPTION
**Pull request by vuonojenmustaturska**
## About the PR 
This adds a system for "mob properties", a fast-to-check system useful for many things. Included are a couple of variants:

- Simple properties, which function like multi-source flags (similar to tg traits),
- Sum properties, where the value is the sum of all applied sources
- Max properties, where the value is the largest of all applied sources
- Priority properties, where the value is the same as the source with the largest priority value

A simple property is used to refactor `update_canmove()` a bit. 

## Why's this needed? 
hasStatus loops through status effects, which is fairly inconvenient when a lot of status effects have similar effects on the mob. It would be better if the status effects set a prop like this instead.

### CODE REVIEW

**Comment by pali6** [comment]
I like this a lot. But I still think either making a variant of this or chaing this to separate `value` from its priority would be hella useful. So you can have stuff like, uh, effects overriding mob's blood color or something. It's 2 AM and my brain refuses to come up with better examples, ok?

**Comment by vuonojenmustaturska** [comment]
Could split this into multiple defines:
have `APPLY_MOB_PROPERTY_MAX(target, property, source, value)` for things where the value is the maximum, and `APPLY_MOB_PROPERTY_SIMPLE(target, property, source, value)` for things where the property functions as a flag, and `APPLY_MOB_PROPERTY_PRIORITY(target, property, source, value, priority)` where there is a separate priority value which determines which value becomes active.

Could add some proc wrappers that handle setting this so that people don't get tripped from using the wrong define on the wrong property, could also have the property's type be part of the property name define.

**Comment by Tarmunora** [comment]
Maybe have a define for getting the sum value of a property for stuff that you want stack on top of itself, like partial resistances?

**Comment by ZeWaka** [comment]
Looks pretty good. I do like the above comments though.

**Comment by pali6** [comment]
For the sum value I'd just have a different variant of the procs (one that'd do `+=` on ADD and `-=` on REMOVE).

**Comment by pali6** [comment]
With custom operators this additive variant could potentially be used for more complex composition of datums too.

**Comment by pali6** [comment]
We could store the property composition method and other related info in the property defintion itself through some shameless abuse of macros and ternary operators. An example of that:
```
#define TRIPLE(x, y, z) x:y):z
#define TRIPLE_1ST(t) (1?(1?t)
#define TRIPLE_2ND(t) (1?(0?t)
#define TRIPLE_3RD(t) (0?(0?t)

#define TRIPLE_EXAMPLE TRIPLE(1, 2, 3)
#define PRINT_TRIPLE(t) world.log << "[TRIPLE_1ST(t)] [TRIPLE_2ND(t)] [TRIPLE_3RD(t)]"
```

**Comment by vuonojenmustaturska** [comment]
Added some helper procs and more documentation, for better ergonomics.

**Comment by pali6** [comment]
Zero overhead idea for storing multiple values in a macro:
```
#define EXAMPLE_TRIPLE(x) x("one", "two", "three")

#define TRIPLE_GET_1ST(a, b, c) a
#define TRIPLE_GET_2ND(a, b, c) b
#define TRIPLE_GET_3RD(a, b, c) c

#define TRIPLE_1ST(x) x(TRIPLE_GET_1ST)
#define TRIPLE_2ND(x) x(TRIPLE_GET_2ND)
#define TRIPLE_3RD(x) x(TRIPLE_GET_3RD)

#define PRINT_TRIPLE(t) world.log << "[TRIPLE_1ST(t)] [TRIPLE_2ND(t)] [TRIPLE_3RD(t)]"

PRINT_TRIPLE(EXAMPLE_TRIPLE)
```
Instead of the string values above those can also be *other macros* which is what would be helpful here so you can set the type of the property in its definition.

**Comment by ZeWaka** [comment]
For priority properties, we could use named arguments just to make clear what the two numbers are, perhaps

**Comment by vuonojenmustaturska** [comment]
Latest commit has defines reworked to be in line with what pali suggested, this allows for the accessor macros to be defined with the magic string which should help ergonomics quite a bit

File: code/modules/antagonists/changeling/abilities/regeneration.dm [hunk 1]
```diff
@@ -33,6 +33,7 @@
 				implants += I
 
 			H.in_fakedeath = 1
+			APPLY_MOB_PROPERTY(C, PROP_CANTMOVE, "fakedeath")
```
Comment (ZeWaka) [inline code/modules/antagonists/changeling/abilities/regeneration.dm:36]: Should this and regeneration.dm:73 be using the same index or whatever? One uses src.type, one uses "fakedeath"

Comment (vuonojenmustaturska) [reply code/modules/antagonists/changeling/abilities/regeneration.dm:36]: oh yes definitely

**Comment by pali6** [comment]
A thought for later: A fourth (optional) element to the property definition which is a macro that gets called whenever the current value of the property changes. (E.g. Whenever property nightvision changes you'd update current mob's centerlight).

**Pull request merged**